docs/RFCS/error_handling.md

```diff
@@ -0,0 +1,97 @@
+- Feature Name: Error Handling
+- Status: draft
+- Start Date: 2017-05-12
+- Authors: knz, dt
+- RFC PR:
+- Cockroach Issue:  #5452
+
+# Summary
+
+Establish patterns and helper utilities to ensure consistent handling and
+presentation of errors.
+
+# Background
+
+Much of our code existing uses the ubiquitous `if err != nil { return err }`
+pattern, sometimes with the addition of context via an `errors.Wrapf` call.
+These errors can bubble up to our SQL connection handler or to a callsite that
+simply passes them to `panic`. Many callsites to `panic` or `log.Fatal` construct
+a string or error object containing both a statement of what went wrong as well
+the value(s) in question, e.g. `missing value in column %s` or
+`unexpected value: raw_bytes:"..."`.
+
+# Motivation
+
+Bubbling the original error back up, adding as much context as possible along
+the way, makes it easier to see what went wrong and thus figure out how to fix
+it. Unfortunately it also means that when our connection handler (pgwire) gets
+an error, or our panic handler recovers a crash, we can make very limited
+assumptions about what is actually in the error, and thus whether it is suitable
+for presentation to the user, inclusion in a diagnostic crash report, etc.
+
+Presenting users (SQL clients) with the details of some internal (e.g. KV) error
+is poor UX: we should strive to provide useful and actionable errors that the
+user can do something about. Internal errors should be logged internally (to
+our log files) where the administrator can find and troubleshoot then.
+
+Error messages that contain potentially sensitive user data cannot be included
+in crash reports, and at some point in the future, if/when we add support for
+encrypting stores, we may want an option to avoid inclusion in (unencrypted) log
+files as well.
+
+# Detailed Design
+
+## Presentation to Users (i.e. errors to SQL clients)
+TL;DR: sql executor will only render and return `pgerror.Error`-derived errors.
+
+While implementing SQL execution, we must consciously choose when constructing
+an error whether it is intended for presentation to a user, and use `pgerror`
+helpers when it is to correctly indicate that, along with its matching error
+code.
+
+Errors *not* intended for user consumption should be wrapped in some sort of
+`InternalError` which, when presented to the user, simply renders something like
+`internal error (logged as <nodeid:id>)` while logged the wrapped error to its
+logs with the unique ID, such that a DBA with log access, presented with the
```

Comment:
Is that immediately useful? My worry is that it makes everyone's life harder and disincentivizes users who run into internal errors from reporting them (or, reporting them regardless and inflicting several back-and-forths on everyone involved). If you think this feature is indispensable, perhaps a cluster setting with default to spilling the beans?